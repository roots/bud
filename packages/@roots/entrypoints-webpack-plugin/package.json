{
  "name": "@roots/entrypoints-webpack-plugin",
  "description": "Manifest with assets grouped by entrypoint",
  "version": "4.7.0-9",
  "repository": {
    "type": "git",
    "url": "https://github.com/roots/bud.git",
    "directory": "packages/@roots/entrypoints-webpack-plugin"
  },
  "contributors": [
    {
      "name": "Kelly Mears",
      "url": "https://github.com/kellymears"
    }
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/roots/bud/issues"
  },
  "funding": {
    "type": "github sponsors",
    "url": "https://github.com/sponsors/roots"
  },
  "keywords": [
    "webpack",
    "asset-manifest"
  ],
  "engines": {
    "node": ">=14"
  },
  "files": [
    "docs/",
    "lib/",
    "types/",
    "manifest.yml"
  ],
  "main": "./lib/cjs/index.js",
  "module": "./lib/esm/index.js",
  "types": "./types/index.d.ts",
  "exports": {
    "require": "./lib/cjs/index.js",
    "import": "./lib/esm/index.js"
  },
  "scripts": {
    "build": "yarn g:build",
    "build:cjs": "yarn g:build:cjs",
    "build:esm": "yarn g:build:esm",
    "profile": "yarn g:profile",
    "profile:cjs": "yarn g:profile:cjs",
    "profile:esm": "yarn g:profile:esm",
    "lint": "yarn g:lint",
    "lint:eslint": "yarn g:lint:eslint",
    "lint:prettier": "yarn g:lint:prettier",
    "clean": "yarn g:clean",
    "clean:lib": "yarn g:clean:lib",
    "clean:modules": "yarn g:clean:modules",
    "clean:types": "yarn g:clean:types",
    "pkg": "yarn g:pkg",
    "pkg:check": "yarn g:pkg:check",
    "pkg:format": "yarn g:pkg:format"
  },
  "devDependencies": {
    "@types/node": "^15.12.5"
  },
  "dependencies": {
    "autobind-decorator": "^2.4.0",
    "lodash": "^4.17.21",
    "webpack": "^5.44.0"
  },
  "stableVersion": "4.6.0"
}
