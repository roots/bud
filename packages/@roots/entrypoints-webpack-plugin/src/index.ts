import Webpack from 'webpack'
import {boundMethod as bind} from 'autobind-decorator'

/**
 * Entrypoints Webpack Plugin
 */
export class Plugin {
  public plugin = {
    name: 'EntrypointsManifestPlugin',
    stage: Infinity,
  }

  public name: string = 'entrypoints.json'

  public webpack: {
    compiler: Webpack.Compiler
    compilation: Webpack.Compilation
  } = {
    compiler: null,
    compilation: null,
  }

  public assets: {[key: string]: {[key: string]: string[]}}

  @bind
  apply(compiler: Webpack.Compiler): void {
    this.assets = {}

    this.webpack.compiler = compiler

    this.webpack.compiler.hooks.thisCompilation.tap(
      this.plugin,
      this.compilation,
    )
  }

  @bind
  public compilation(compilation: Webpack.Compilation) {
    this.webpack.compilation = compilation

    this.webpack.compilation.hooks.processAssets.tap(
      {...this.plugin, additionalAssets: true},
      this.processAssets,
    )
  }

  @bind
  public processAssets() {
    this.webpack.compilation.entrypoints.forEach(entry => {
      this.getEntrypointFiles(entry).map(file => {
        !file.includes('hot-update') &&
          this.addToManifest(entry.name, file)
      })
    })

    this.webpack.compilation.assets[
      this.name
    ] = new Webpack.sources.RawSource(
      JSON.stringify({...this.assets}),
      true,
    )
  }

  @bind
  public addToManifest(entry, file) {
    const type = file.split('.').pop()

    this.assets[entry] = {
      ...(this.assets[entry] ?? {}),
      [type]:
        this.assets[entry] &&
        this.assets[entry][type] &&
        !this.assets[entry][type].includes(file)
          ? [...this.assets[entry][type], file]
          : [file],
    }
  }

  @bind
  public publicPath(path: string) {
    return `${this.webpack.compiler.options.output.publicPath}${path}`
  }

  @bind
  public getEntrypointFiles(entry: {
    chunks: Webpack.Chunk[]
    origins: any[]
  }): string[] {
    const files = []
    for (const chunk of entry.chunks) {
      Array.from(chunk.files).map(file => {
        files.push(file)
      })
    }

    return files
  }
}

/**
 * Schema for manifest entry
 */
export type EntrySchema = {
  [key: string]: {
    [type: string]: {
      [handle: string]: string
    }
  }
}

/**
 * Constructor params
 */
export type Options = {
  name?: string
  writeToFileEmit?: boolean
  publicPath?: string
  outputPath?: string
}
