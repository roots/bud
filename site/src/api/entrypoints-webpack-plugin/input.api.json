{
  "metadata": {
    "toolPackage": "@microsoft/api-extractor",
    "toolVersion": "7.18.19",
    "schemaVersion": 1004,
    "oldestForwardsCompatibleVersion": 1001,
    "tsdocConfig": {
      "$schema": "https://developer.microsoft.com/json-schemas/tsdoc/v0/tsdoc.schema.json",
      "noStandardTags": true,
      "tagDefinitions": [
        {
          "tagName": "@alpha",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@beta",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@defaultValue",
          "syntaxKind": "block"
        },
        {
          "tagName": "@decorator",
          "syntaxKind": "block",
          "allowMultiple": true
        },
        {
          "tagName": "@deprecated",
          "syntaxKind": "block"
        },
        {
          "tagName": "@eventProperty",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@example",
          "syntaxKind": "block",
          "allowMultiple": true
        },
        {
          "tagName": "@experimental",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@inheritDoc",
          "syntaxKind": "inline"
        },
        {
          "tagName": "@internal",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@label",
          "syntaxKind": "inline"
        },
        {
          "tagName": "@link",
          "syntaxKind": "inline",
          "allowMultiple": true
        },
        {
          "tagName": "@override",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@packageDocumentation",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@param",
          "syntaxKind": "block",
          "allowMultiple": true
        },
        {
          "tagName": "@privateRemarks",
          "syntaxKind": "block"
        },
        {
          "tagName": "@public",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@readonly",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@remarks",
          "syntaxKind": "block"
        },
        {
          "tagName": "@returns",
          "syntaxKind": "block"
        },
        {
          "tagName": "@sealed",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@see",
          "syntaxKind": "block"
        },
        {
          "tagName": "@throws",
          "syntaxKind": "block",
          "allowMultiple": true
        },
        {
          "tagName": "@typeParam",
          "syntaxKind": "block",
          "allowMultiple": true
        },
        {
          "tagName": "@virtual",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@betaDocumentation",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@internalRemarks",
          "syntaxKind": "block"
        },
        {
          "tagName": "@preapproved",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@container",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@core",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@extension",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@config",
          "syntaxKind": "modifier"
        },
        {
          "tagName": "@hook",
          "syntaxKind": "block",
          "allowMultiple": true
        },
        {
          "tagName": "@filter",
          "syntaxKind": "block",
          "allowMultiple": true
        }
      ],
      "supportForTags": {
        "@alpha": true,
        "@beta": true,
        "@defaultValue": true,
        "@decorator": true,
        "@deprecated": true,
        "@eventProperty": true,
        "@example": true,
        "@experimental": true,
        "@inheritDoc": true,
        "@internal": true,
        "@label": true,
        "@link": true,
        "@override": true,
        "@packageDocumentation": true,
        "@param": true,
        "@privateRemarks": true,
        "@public": true,
        "@readonly": true,
        "@remarks": true,
        "@returns": true,
        "@sealed": true,
        "@see": true,
        "@throws": true,
        "@typeParam": true,
        "@virtual": true,
        "@betaDocumentation": true,
        "@internalRemarks": true,
        "@preapproved": true,
        "@container": true,
        "@core": true,
        "@extension": true,
        "@config": true,
        "@hook": true,
        "@filter": true
      }
    }
  },
  "kind": "Package",
  "canonicalReference": "@roots/entrypoints-webpack-plugin!",
  "docComment": "/**\n * Webpack plugin which produces an `entrypoints.json` artifact organizing compiled assets by entrypoint and filetype\n *\n * @packageDocumentation\n */\n",
  "name": "@roots/entrypoints-webpack-plugin",
  "members": [
    {
      "kind": "EntryPoint",
      "canonicalReference": "@roots/entrypoints-webpack-plugin!",
      "name": "",
      "members": [
        {
          "kind": "Class",
          "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin:class",
          "docComment": "/**\n * Produces `entrypoints.json` artifact with compiled assets broken down by entrypoint and then filetype.\n *\n * @example\n * ```\n * import {EntrypointsWebpackPlugin} from '@roots/entrypoints-webpack-plugin'\n *\n * const config = {\n *   plugins: [new EntrypointsWebpackPlugin()]\n * }\n * ```\n *\n * @sealed\n */\n",
          "excerptTokens": [
            {
              "kind": "Content",
              "text": "export declare class EntrypointsWebpackPlugin implements "
            },
            {
              "kind": "Reference",
              "text": "EntrypointsPlugin",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!~EntrypointsPlugin:interface"
            },
            {
              "kind": "Content",
              "text": " "
            }
          ],
          "releaseTag": "Public",
          "name": "EntrypointsWebpackPlugin",
          "members": [
            {
              "kind": "Constructor",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin:constructor(1)",
              "docComment": "/**\n * Class constructor\n *\n * @public\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "constructor(options?: "
                },
                {
                  "kind": "Reference",
                  "text": "Options",
                  "canonicalReference": "@roots/entrypoints-webpack-plugin!~Options:interface"
                },
                {
                  "kind": "Content",
                  "text": ");"
                }
              ],
              "releaseTag": "Public",
              "overloadIndex": 1,
              "parameters": [
                {
                  "parameterName": "options",
                  "parameterTypeTokenRange": {
                    "startIndex": 1,
                    "endIndex": 2
                  }
                }
              ]
            },
            {
              "kind": "Method",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin#addToManifest:member(1)",
              "docComment": "/**\n * Adds an entry to the manifest\n *\n * @decorator\n *\n * `@bind`\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "addToManifest({ entry, file }: "
                },
                {
                  "kind": "Content",
                  "text": "{\n        entry: any;\n        file: any;\n    }"
                },
                {
                  "kind": "Content",
                  "text": "): "
                },
                {
                  "kind": "Content",
                  "text": "void"
                },
                {
                  "kind": "Content",
                  "text": ";"
                }
              ],
              "isOptional": false,
              "isStatic": false,
              "returnTypeTokenRange": {
                "startIndex": 3,
                "endIndex": 4
              },
              "releaseTag": "Public",
              "overloadIndex": 1,
              "parameters": [
                {
                  "parameterName": "{ entry, file }",
                  "parameterTypeTokenRange": {
                    "startIndex": 1,
                    "endIndex": 2
                  }
                }
              ],
              "name": "addToManifest"
            },
            {
              "kind": "Method",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin#apply:member(1)",
              "docComment": "/**\n * Webpack plugin API's `apply` hook\n *\n * @decorator\n *\n * `@bind`\n *\n * @public\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "apply(compiler: "
                },
                {
                  "kind": "Reference",
                  "text": "Webpack.Compiler",
                  "canonicalReference": "webpack!~Compiler:class"
                },
                {
                  "kind": "Content",
                  "text": "): "
                },
                {
                  "kind": "Content",
                  "text": "void"
                },
                {
                  "kind": "Content",
                  "text": ";"
                }
              ],
              "isOptional": false,
              "isStatic": false,
              "returnTypeTokenRange": {
                "startIndex": 3,
                "endIndex": 4
              },
              "releaseTag": "Public",
              "overloadIndex": 1,
              "parameters": [
                {
                  "parameterName": "compiler",
                  "parameterTypeTokenRange": {
                    "startIndex": 1,
                    "endIndex": 2
                  }
                }
              ],
              "name": "apply"
            },
            {
              "kind": "Property",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin#assets:member",
              "docComment": "/**\n * Collected assets\n *\n * @public\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "assets: "
                },
                {
                  "kind": "Reference",
                  "text": "Entry",
                  "canonicalReference": "@roots/entrypoints-webpack-plugin!~Entry:interface"
                },
                {
                  "kind": "Content",
                  "text": ";"
                }
              ],
              "isOptional": false,
              "releaseTag": "Public",
              "name": "assets",
              "propertyTypeTokenRange": {
                "startIndex": 1,
                "endIndex": 2
              },
              "isStatic": false
            },
            {
              "kind": "Property",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin#compilation:member",
              "docComment": "/**\n * Webpack compilation instance\n *\n * @public\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "compilation: "
                },
                {
                  "kind": "Reference",
                  "text": "Webpack.Compilation",
                  "canonicalReference": "webpack!~Compilation:class"
                },
                {
                  "kind": "Content",
                  "text": ";"
                }
              ],
              "isOptional": false,
              "releaseTag": "Public",
              "name": "compilation",
              "propertyTypeTokenRange": {
                "startIndex": 1,
                "endIndex": 2
              },
              "isStatic": false
            },
            {
              "kind": "Property",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin#compiler:member",
              "docComment": "/**\n * Webpack compiler instance\n *\n * @public\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "compiler: "
                },
                {
                  "kind": "Reference",
                  "text": "Webpack.Compiler",
                  "canonicalReference": "webpack!~Compiler:class"
                },
                {
                  "kind": "Content",
                  "text": ";"
                }
              ],
              "isOptional": false,
              "releaseTag": "Public",
              "name": "compiler",
              "propertyTypeTokenRange": {
                "startIndex": 1,
                "endIndex": 2
              },
              "isStatic": false
            },
            {
              "kind": "Method",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin#getEntrypointFiles:member(1)",
              "docComment": "/**\n * Get assets from an entrypoint\n *\n * @decorator\n *\n * `@bind`\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "getEntrypointFiles(entry: "
                },
                {
                  "kind": "Content",
                  "text": "{\n        chunks: "
                },
                {
                  "kind": "Reference",
                  "text": "Webpack.Chunk",
                  "canonicalReference": "webpack!~Chunk:class"
                },
                {
                  "kind": "Content",
                  "text": "[];\n    }"
                },
                {
                  "kind": "Content",
                  "text": "): "
                },
                {
                  "kind": "Content",
                  "text": "string[]"
                },
                {
                  "kind": "Content",
                  "text": ";"
                }
              ],
              "isOptional": false,
              "isStatic": false,
              "returnTypeTokenRange": {
                "startIndex": 5,
                "endIndex": 6
              },
              "releaseTag": "Public",
              "overloadIndex": 1,
              "parameters": [
                {
                  "parameterName": "entry",
                  "parameterTypeTokenRange": {
                    "startIndex": 1,
                    "endIndex": 4
                  }
                }
              ],
              "name": "getEntrypointFiles"
            },
            {
              "kind": "Property",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin#name:member",
              "docComment": "/**\n * Artifact filename\n *\n * @public\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "name: "
                },
                {
                  "kind": "Content",
                  "text": "string"
                },
                {
                  "kind": "Content",
                  "text": ";"
                }
              ],
              "isOptional": false,
              "releaseTag": "Public",
              "name": "name",
              "propertyTypeTokenRange": {
                "startIndex": 1,
                "endIndex": 2
              },
              "isStatic": false
            },
            {
              "kind": "Property",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin#plugin:member",
              "docComment": "/**\n * Plugin compiler ident\n *\n * @public\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "protected plugin: "
                },
                {
                  "kind": "Content",
                  "text": "{\n        name: string;\n        stage: number;\n    }"
                },
                {
                  "kind": "Content",
                  "text": ";"
                }
              ],
              "isOptional": false,
              "releaseTag": "Public",
              "name": "plugin",
              "propertyTypeTokenRange": {
                "startIndex": 1,
                "endIndex": 2
              },
              "isStatic": false
            },
            {
              "kind": "Method",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin#processAssets:member(1)",
              "docComment": "/**\n * Runs through each entrypoint entry and adds to the manifest\n *\n * @decorator\n *\n * `@bind`\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "processAssets(): "
                },
                {
                  "kind": "Content",
                  "text": "void"
                },
                {
                  "kind": "Content",
                  "text": ";"
                }
              ],
              "isOptional": false,
              "isStatic": false,
              "returnTypeTokenRange": {
                "startIndex": 1,
                "endIndex": 2
              },
              "releaseTag": "Public",
              "overloadIndex": 1,
              "parameters": [],
              "name": "processAssets"
            },
            {
              "kind": "Property",
              "canonicalReference": "@roots/entrypoints-webpack-plugin!EntrypointsWebpackPlugin#publicPath:member",
              "docComment": "/**\n * Project publicPath\n *\n * @public\n */\n",
              "excerptTokens": [
                {
                  "kind": "Content",
                  "text": "publicPath: "
                },
                {
                  "kind": "Content",
                  "text": "string"
                },
                {
                  "kind": "Content",
                  "text": ";"
                }
              ],
              "isOptional": false,
              "releaseTag": "Public",
              "name": "publicPath",
              "propertyTypeTokenRange": {
                "startIndex": 1,
                "endIndex": 2
              },
              "isStatic": false
            }
          ],
          "implementsTokenRanges": [
            {
              "startIndex": 1,
              "endIndex": 3
            }
          ]
        }
      ]
    }
  ]
}
