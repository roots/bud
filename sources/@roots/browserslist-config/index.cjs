/**
 * Browserslist default
 */
module.exports = [
  `> 0.5%`,
  `last 2 versions`,
  `not dead`,
  `not op_mini all`,
  `supports es6-module-dynamic-import`,
]
