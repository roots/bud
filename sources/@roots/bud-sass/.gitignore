test/fixtures/issue-1798/.budfiles
test/fixtures/issue-1798/dist
