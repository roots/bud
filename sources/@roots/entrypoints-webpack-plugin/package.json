{
  "name": "@roots/entrypoints-webpack-plugin",
  "description": "Manifest with assets grouped by entrypoint",
  "version": "0.0.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/roots/bud.git",
    "directory": "sources/@roots/entrypoints-webpack-plugin"
  },
  "contributors": [
    {
      "name": "kellymears",
      "url": "https://github.com/kellymears"
    },
    {
      "name": "QWp6t",
      "url": "https://github.com/QWp6t"
    }
  ],
  "license": "MIT",
  "bugs": {
    "url": "https://github.com/roots/bud/issues"
  },
  "funding": {
    "type": "github sponsors",
    "url": "https://github.com/sponsors/roots"
  },
  "keywords": [
    "webpack",
    "asset-manifest"
  ],
  "engines": {
    "node": ">=16"
  },
  "files": [
    "lib/"
  ],
  "type": "module",
  "module": "./lib/index.js",
  "types": "./lib/index.d.ts",
  "exports": {
    ".": "./lib/index.js"
  },
  "typesVersions": {
    "*": {
      ".": [
        "./lib/index.d.ts"
      ]
    }
  },
  "devDependencies": {
    "@skypack/package-check": "0.2.2",
    "@types/jest": "27.5.1",
    "@types/lodash-es": "^4",
    "@types/node": "16.11.36",
    "tslib": "2.4.0",
    "webpack": "5.72.1"
  },
  "dependencies": {
    "helpful-decorators": "2.1.0",
    "lodash-es": "4.17.21"
  },
  "peerDependencies": {
    "webpack": ">=5"
  }
}
