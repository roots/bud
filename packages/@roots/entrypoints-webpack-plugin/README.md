<p align="center">
  <img alt="Bud" src="https://cdn.roots.io/app/uploads/logo-bud.svg" height="100">
</p>

<p align="center">
  <img alt="MIT License" src="https://img.shields.io/github/license/roots/bud?color=%23525ddc&style=flat-square">
  <a href="https://www.npmjs.com/package/@roots/entrypoints-webpack-plugin">
    <img src="https://img.shields.io/npm/v/@roots/entrypoints-webpack-plugin.svg?color=%23525ddc&style=flat-square" />
  </a>
  <a href="https://codeclimate.com/github/roots/bud-support/maintainability">
    <img src="https://img.shields.io/codeclimate/maintainability/roots/bud-support?color=%23525ddc&style=flat-square" />
  </a>
  <a href="https://github.com/roots/bud/actions/workflows/build">
    <img src="https://github.com/roots/bud/actions/workflows/build.yml/badge.svg" />
  </a>
  <a href="Typescript" src="https://github.com/roots/bud/tree/stable/typings">
    <img src="https://img.shields.io/badge/typings-%40roots%2Fbud--typings-%23525ddc" />
  </a>
  <a href="https://twitter.com/rootswp">
    <img alt="Follow Roots" src="https://img.shields.io/twitter/follow/rootswp.svg?color=%23525ddc&style=flat-square" />
  </a>
</p>

<h1 align="center">
  <strong>@roots/entrypoints-webpack-plugin</strong>
</h1>

> undefined

- [Overview](#overview)
- [Installation](#installation)
- [Usage](#usage)
- [Output](#output)
- [WordPress example implementation](#wordpress-example-implementation)
- [Contributing](#contributing)
- [Bud sponsors](#bud-sponsors)
- [Community](#community)

## Overview

Intended for use in environments where a separate, server-side process handles enqueuing assets for the client (see: [**@wordpress/wordpress**](https://github.com/wordpress/wordpress) or [**@laravel/laravel**](https://github.com/laravel/laravel)).

[**@roots/entrypoints-webpack-plugin**](https://github.com/roots/bud/tree/stable/packages/@roots/entrypoints-webpack-plugin) is a vanilla Webpack plugin and can be used outside of [**@roots/bud**](https://github.com/roots/bud/tree/stable/packages/@roots/bud) projects.

If you want to use it with [**@roots/bud**](https://github.com/roots/bud/tree/stable/packages/@roots/bud), you should instead use [**@roots/bud-entrypoints-webpack-plugin**](https://github.com/roots/bud/tree/stable/packages/@roots/bud-entrypoints-webpack-plugin), which wraps this plugin.

## Installation

```sh
yarn add @roots/entrypoints-webpack-plugin --dev
```

## Usage

```js
import { Plugin } from "@roots/entrypoints-webpack-plugin";

webpack({
  plugins: [new Plugin()],
});
```

## Output

Assets are separated by entrypoint name, and then by filetype. The `runtimeChunk`, if used, will be the first item in the entrypoint js assets array.

```json
{
  "app": {
    "js": ["runtime.9ffe4f626b6ffca53bd1.js", "app.d4e917b74a97bb0fe9ad.js"],
    "css": ["app.bcaf7e71ef4356832914.css"]
  },
  "editor": {
    "js": ["runtime.9ffe4f626b6ffca53bd1.js", "editor.a22827c63d6d9e772ea1.js"],
    "css": ["editor.ec414c4a744ec7bcb597.css"]
  },
  "customizer": {
    "js": [
      "runtime.9ffe4f626b6ffca53bd1.js",
      "customizer.8d9da55b6e4e522dfe4b.js"
    ]
  }
}
```

## WordPress example implementation

This is a bad implementation but [**@roots/sage**](https://github.com/roots/sage) comes with an implementation that is quite good.

```php
if (!$path = realpath(get_theme_file_path('dist/entrypoints.json')) {
    throw new \WP_Error('Run yarn build');
}

Collection::make(json_decode(file_get_contents($path)))->each(
    function ($entrypoint, $name) {
        Collection::make($entrypoint->js)->each(
            function ($asset) use ($name) {
                wp_enqueue_script(...[
                    "$name/$asset",
                    get_template_directory_uri("dist/$asset"),
                    [],
                    null,
                ]);
            }
        )
    }
);
```

## Contributing

Contributions are welcome from everyone.

We have [contributing guidelines](https://github.com/roots/guidelines/blob/master/CONTRIBUTING.md) to help you get started.

## Bud sponsors

Help support our open-source development efforts by [becoming a patron](https://www.patreon.com/rootsdev).

<a href="https://kinsta.com/?kaid=OFDHAJIXUDIV">
  <img src="https://cdn.roots.io/app/uploads/kinsta.svg" alt="Kinsta" width="200" height="150">
</a>
<a href="https://k-m.com/">
  <img src="https://cdn.roots.io/app/uploads/km-digital.svg" alt="KM Digital" width="200" height="150">
</a>
<a href="https://carrot.com/">
  <img src="https://cdn.roots.io/app/uploads/carrot.svg" alt="Carrot" width="200" height="150">
</a>
<a href="https://www.c21redwood.com/">
  <img src="https://cdn.roots.io/app/uploads/c21redwood.svg" alt="C21 Redwood Realty" width="200" height="150">
</a>
<a href="https://wordpress.com/">
  <img src="https://cdn.roots.io/app/uploads/wordpress.svg" alt="WordPress.com" width="200" height="150">
</a>
<a href="https://icons8.com/">
  <img src="https://cdn.roots.io/app/uploads/icons8.svg" alt="Icons8" width="200" height="150">
</a>
<a href="https://www.harnessup.com/">
  <img src="https://cdn.roots.io/app/uploads/harness-software.svg" alt="Harness Software" width="200" height="150">
</a>
<a href="https://www.codersclan.com/">
  <img src="https://cdn.roots.io/app/uploads/coders-clan.svg" alt="Coders Clan" width="200" height="150">
</a>
<a href="https://generodigital.com/">
  <img src="https://cdn.roots.io/app/uploads/genero.svg" alt="Genero" width="200" height="150">
</a>
<a href="https://motto.ca/roots">
  <img src="https://cdn.roots.io/app/uploads/motto.svg" alt="Motto" width="200" height="150">
</a>

## Community

Keep track of development and community news.

- Participate on the [Roots Discourse](https://discourse.roots.io/)
- Follow [@rootswp on Twitter](https://twitter.com/rootswp)
- Read and subscribe to the [Roots Blog](https://roots.io/blog/)
- Subscribe to the [Roots Newsletter](https://roots.io/subscribe/)
- Listen to the [Roots Radio podcast](https://roots.io/podcast/)
