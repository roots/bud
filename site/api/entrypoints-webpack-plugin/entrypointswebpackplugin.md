---
id: entrypoints-webpack-plugin.entrypointswebpackplugin
title: entrypointswebpackplugin class
sidebar_label: entrypointswebpackplugin class
hide_title: true
sidebar: "api"
slug: entrypointswebpackplugin
---

<!-- Do not edit this file. It is automatically generated by API Documenter. -->

## EntrypointsWebpackPlugin class

Produces `entrypoints.json` artifact with compiled assets broken down by entrypoint and then filetype.

Signature:

```typescript
/** @sealed */
export declare class EntrypointsWebpackPlugin implements EntrypointsPlugin
```

Implements: EntrypointsPlugin

## Example

```
import {EntrypointsWebpackPlugin} from '@roots/entrypoints-webpack-plugin'

const config = {
  plugins: [new EntrypointsWebpackPlugin()]
}
```

## Constructors

| Constructor                                                                                    | Modifiers | Description       |
| ---------------------------------------------------------------------------------------------- | --------- | ----------------- |
| [(constructor)(options)](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/constructor) |           | Class constructor |

## Properties

| Property                                                                            | Modifiers | Type                             | Description                                               |
| ----------------------------------------------------------------------------------- | --------- | -------------------------------- | --------------------------------------------------------- |
| [assets](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/assets)           |           | Entry                            | Collected assets                                          |
| [compilation](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/compilation) |           | Webpack.Compilation              | Webpack compilation instance                              |
| [compiler](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/compiler)       |           | Webpack.Compiler                 | Webpack compiler instance                                 |
| [name](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/name)               |           | string                           | Artifact filename                                         |
| [plugin](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/plugin)           |           | { name: string; stage: number; } | Plugin compiler ident                                     |
| [publicPath](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/publicpath)   |           | string                           | Project publicPath                                        |
| [type](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/type)               |           | 'array' &#124; 'object'          | Entrypoints type.Object type will key files by chunk name |

## Methods

| Method                                                                                                        | Modifiers | Description                                                 |
| ------------------------------------------------------------------------------------------------------------- | --------- | ----------------------------------------------------------- |
| [addToManifest({ key, entry, file })](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/addtomanifest) |           | Adds an entry to the manifest                               |
| [apply(compiler)](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/apply)                             |           | Webpack plugin API's <code>apply</code> hook                |
| [getEntrypointFiles(entry)](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/getentrypointfiles)      |           | Get assets from an entrypoint                               |
| [processAssets()](/api/entrypoints-webpack-plugin/entrypointswebpackplugin/processassets)                     |           | Runs through each entrypoint entry and adds to the manifest |
