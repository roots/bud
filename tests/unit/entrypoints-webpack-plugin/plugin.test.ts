import {EntrypointsWebpackPlugin} from '@roots/entrypoints-webpack-plugin'
import Webpack from 'webpack'

describe('entrypoints.json', () => {
  it('should get chunk file list', () => {
    const entrypoints = new EntrypointsWebpackPlugin({
      publicPath: '/public/',
    })
    const chonk = new Webpack.Chunk()
    chonk.files = new Set(['foo.js', 'bar.js'])
    const files = entrypoints.getEntrypointFiles({
      chunks: [chonk],
    })

    expect(files).toMatchSnapshot([{file: 'foo.js'}, {file: 'bar.js'}])
  })

  it('should create manifest object', () => {
    const entrypoints = new EntrypointsWebpackPlugin({
      publicPath: '/public/',
    })

    entrypoints.assets = {}

    entrypoints.addToManifest({
      entry: 'app',
      file: 'runtime.js',
    })
    entrypoints.addToManifest({
      entry: 'app',
      file: 'app.js',
    })
    entrypoints.addToManifest({
      entry: 'app',
      file: 'app.css',
    })
    entrypoints.addToManifest({
      entry: 'app',
      file: 'vendor/foobar.js',
    })

    expect(entrypoints.assets).toEqual({
      app: {
        js: [
          '/public/runtime.js',
          '/public/app.js',
          '/public/vendor/foobar.js',
        ],
        css: ['/public/app.css'],
      },
    })
  })

  it('should create manifest object with keyed file list', () => {
    const entrypoints = new EntrypointsWebpackPlugin({
      publicPath: '/public/',
      type: 'object',
    })

    entrypoints.assets = {}

    entrypoints.addToManifest({
      key: 'runtime',
      entry: 'app',
      file: 'runtime.js',
    })
    entrypoints.addToManifest({
      key: 'app',
      entry: 'app',
      file: 'app.js',
    })
    entrypoints.addToManifest({
      key: 'app',
      entry: 'app',
      file: 'app.css',
    })
    entrypoints.addToManifest({
      key: 'vendor/foobar',
      entry: 'app',
      file: 'vendor/foobar.js',
    })

    expect(entrypoints.assets).toEqual({
      app: {
        js: {
          runtime: '/public/runtime.js',
          app: '/public/app.js',
          'vendor/foobar': '/public/vendor/foobar.js',
        },
        css: {app: '/public/app.css'},
      },
    })
  })
})
