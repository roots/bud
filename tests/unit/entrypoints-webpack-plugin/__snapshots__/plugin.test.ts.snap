// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`entrypoints.json should get chunk file list 1`] = `
[
  {
    "file": "foo.js",
    "key": undefined,
  },
  {
    "file": "bar.js",
    "key": undefined,
  },
]
`;
